File: example/r1/high_level/r1_tool.cpp
// ============================================================================
// r1_tool.cpp
// R1 官方 SDK 全部服务命令行测试工具（二次开发前验证用）。
//
// 覆盖本 SDK 为 R1 提供的所有高层服务：
//   loco   -> r1::LocoClient            (DDS 服务 "sport")
//   msc    -> b2::MotionSwitcherClient   ("motion_switcher")
//   state  -> b2::RobotStateClient       ("robot_state")
//   audio  -> r1::AudioClient            ("voice")
//   config -> b2::ConfigClient           ("config")
//   lowstate -> 订阅 "rt/lowstate"（只读，安全）
//
// 用法：
//   r1_tool <network_interface> <group> <action> [args...]
//   r1_tool help
//
// 例：
//   r1_tool eth0 loco get-fsm
//   r1_tool eth0 loco stand
//   r1_tool eth0 msc check
//   r1_tool eth0 state services
//   r1_tool eth0 audio tts "hello"
//   r1_tool eth0 lowstate dump 3
// ============================================================================

#include <atomic>
#include <chrono>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <map>
#include <mutex>
#include <string>
#include <thread>
#include <vector>

#include "unitree/robot/channel/channel_factory.hpp"
#include "unitree/robot/channel/channel_subscriber.hpp"
#include "unitree/idl/hg/LowState_.hpp"
#include "unitree/idl/ros2/String_.hpp"

#include "unitree/robot/r1/loco/r1_loco_client.hpp"
#include "unitree/robot/r1/audio/audio_client.hpp"
#include "unitree/robot/b2/motion_switcher/motion_switcher_client.hpp"
#include "unitree/robot/b2/robot_state/robot_state_client.hpp"
#include "unitree/robot/b2/config/config_client.hpp"

using std::string;
using std::vector;

static void print_usage() {
  printf(
    "R1 官方 SDK 全部服务测试工具\n"
    "\n"
    "用法: r1_tool <network_interface> <group> <action> [args...]\n"
    "      r1_tool help\n"
    "\n"
    "group/action:\n"
    "  loco get-fsm                         查询 FSM ID        (sport 7001)\n"
    "  loco get-fsm-mode                    查询 FSM Mode      (sport 7002)\n"
    "  loco set-fsm <id>                    设置 FSM ID        (sport 7101)\n"
    "  loco damp                            阻尼  fsm 1\n"
    "  loco stand                           站立  fsm 4\n"
    "  loco start                           启动主运控 fsm 811\n"
    "  loco zero-torque                     零力矩 fsm 0 (危险)\n"
    "  loco stop                            停止移动\n"
    "  loco move <vx> <vy> <vyaw> [dur]     速度移动           (sport 7105)\n"
    "  loco speed <mode>                    速度档位           (sport 7107)\n"
    "\n"
    "  msc check                            查询当前运动模式   (motion_switcher 1001)\n"
    "  msc select <name|alias>              选择模式（如 ai）  (1002)\n"
    "  msc release                          释放当前模式       (1003)\n"
    "  msc silent-get                       查询静默状态       (1005)\n"
    "  msc silent-set <0|1>                 设置静默           (1004)\n"
    "\n"
    "  state services                       列出所有服务及状态 (robot_state 1003)\n"
    "  state version                        查询版本           (1006)\n"
    "  state lowpower-get                   低功耗状态         (1005)\n"
    "  state lowpower-set <0|1>             低功耗开关         (1004)\n"
    "  state report <interval_ms> <dur_ms>  设置上报频率       (1002)\n"
    "  state switch <name> <0|1>            服务开关           (1001)\n"
    "\n"
    "  audio tts <text> [speaker]           语音合成           (voice 1001)\n"
    "  audio vol-get                        查询音量           (1005)\n"
    "  audio vol-set <0-100>                设置音量           (1006)\n"
    "  audio led <r> <g> <b>                RGB 灯             (1010)\n"
    "  audio asr [sec]                      监听 ASR 识别结果  (topic rt/audio_msg)\n"
    "\n"
    "  config get <name>                    读取配置           (config 1002)\n"
    "  config set <name> <content>          写配置             (1001)\n"
    "  config del <name>                    删除配置           (1003)\n"
    "  config meta <name>                   配置元信息         (1004)\n"
    "\n"
    "  lowstate dump [sec]                  订阅 rt/lowstate 打印 IMU + 35 电机状态\n"
    "\n"
    "注意：危险动作（zero-torque / move）请确保机器人周围安全、有保护措施。\n");
}

static int need(const vector<string>& a, size_t n, const char* usage) {
  if (a.size() < n) { fprintf(stderr, "参数不足: %s\n", usage); return -1; }
  return 0;
}

// ---------------------------------------------------------------------------
// loco
// ---------------------------------------------------------------------------
static int cmd_loco(const vector<string>& a) {
  if (a.empty()) { fprintf(stderr, "用法: loco <action>\n"); return 1; }
  unitree::robot::r1::LocoClient c;
  c.Init();
  c.SetTimeout(5.0f);
  const string& act = a[0];
  int r = 0, id = 0, mode = 0;

  if (act == "get-fsm") {
    r = c.GetFsmId(id);
    printf("GetFsmId ret=%d fsm_id=%d\n", r, id);
  } else if (act == "get-fsm-mode") {
    r = c.GetFsmMode(mode);
    printf("GetFsmMode ret=%d fsm_mode=%d\n", r, mode);
  } else if (act == "set-fsm") {
    if (need(a, 2, "loco set-fsm <id>")) return 1;
    id = std::atoi(a[1].c_str());
    r = c.SetFsmId(id);
    printf("SetFsmId(%d) ret=%d\n", id, r);
  } else if (act == "damp") {
    r = c.Damp();            printf("Damp ret=%d\n", r);
  } else if (act == "stand") {
    r = c.StandUp();         printf("StandUp ret=%d\n", r);
  } else if (act == "start") {
    r = c.Start();           printf("Start ret=%d\n", r);
  } else if (act == "zero-torque") {
    r = c.ZeroTorque();      printf("ZeroTorque ret=%d\n", r);
  } else if (act == "stop") {
    r = c.StopMove();        printf("StopMove ret=%d\n", r);
  } else if (act == "move") {
    if (need(a, 4, "loco move <vx> <vy> <vyaw> [dur]")) return 1;
    const float vx = std::atof(a[1].c_str());
    const float vy = std::atof(a[2].c_str());
    const float vyaw = std::atof(a[3].c_str());
    const float dur = (a.size() > 4) ? std::atof(a[4].c_str()) : 1.0f;
    r = c.SetVelocity(vx, vy, vyaw, dur);
    printf("SetVelocity(%.3f %.3f %.3f, %.3fs) ret=%d\n", vx, vy, vyaw, dur, r);
  } else if (act == "speed") {
    if (need(a, 2, "loco speed <mode>")) return 1;
    mode = std::atoi(a[1].c_str());
    r = c.SetSpeedMode(mode);
    printf("SetSpeedMode(%d) ret=%d\n", mode, r);
  } else {
    fprintf(stderr, "未知 loco action: %s\n", act.c_str());
    return 1;
  }
  return r == 0 ? 0 : 2;
}

// ---------------------------------------------------------------------------
// msc (motion_switcher)
// ---------------------------------------------------------------------------
static int cmd_msc(const vector<string>& a) {
  if (a.empty()) { fprintf(stderr, "用法: msc <action>\n"); return 1; }
  unitree::robot::b2::MotionSwitcherClient c;
  c.Init();
  c.SetTimeout(5.0f);
  const string& act = a[0];
  int r = 0;

  if (act == "check") {
    string form, name;
    r = c.CheckMode(form, name);
    printf("CheckMode ret=%d form=\"%s\" name=\"%s\"\n", r, form.c_str(), name.c_str());
    if (name.empty()) printf("(当前无激活的运动模式；ai_sport 若在运行此处可能显示为 ai)\n");
  } else if (act == "select") {
    if (need(a, 2, "msc select <name|alias>")) return 1;
    r = c.SelectMode(a[1]);
    printf("SelectMode(%s) ret=%d\n", a[1].c_str(), r);
  } else if (act == "release") {
    r = c.ReleaseMode();
    printf("ReleaseMode ret=%d\n", r);
  } else if (act == "silent-get") {
    bool s = false;
    r = c.GetSilent(s);
    printf("GetSilent ret=%d silent=%d\n", r, s ? 1 : 0);
  } else if (act == "silent-set") {
    if (need(a, 2, "msc silent-set <0|1>")) return 1;
    r = c.SetSilent(std::atoi(a[1].c_str()) != 0);
    printf("SetSilent(%s) ret=%d\n", a[1].c_str(), r);
  } else {
    fprintf(stderr, "未知 msc action: %s\n", act.c_str());
    return 1;
  }
  return r == 0 ? 0 : 2;
}

// ---------------------------------------------------------------------------
// state (robot_state)
// ---------------------------------------------------------------------------
static int cmd_state(const vector<string>& a) {
  if (a.empty()) { fprintf(stderr, "用法: state <action>\n"); return 1; }
  unitree::robot::b2::RobotStateClient c;
  c.Init();
  c.SetTimeout(5.0f);
  const string& act = a[0];
  int r = 0;

  if (act == "services") {
    vector<unitree::robot::b2::ServiceState> list;
    r = c.ServiceList(list);
    printf("ServiceList ret=%d count=%zu\n", r, list.size());
    for (auto& s : list)
      printf("  %-24s status=%d protect=%d\n", s.name.c_str(), s.status, s.protect);
  } else if (act == "version") {
    string pkg;
    std::map<string, string> mods;
    r = c.GetPkgVersion(pkg, mods);
    printf("GetPkgVersion ret=%d packageVersion=%s\n", r, pkg.c_str());
    for (auto& kv : mods) printf("  %-24s %s\n", kv.first.c_str(), kv.second.c_str());
  } else if (act == "lowpower-get") {
    int32_t st = 0;
    r = c.LowPowerStatus(st);
    printf("LowPowerStatus ret=%d status=%d\n", r, st);
  } else if (act == "lowpower-set") {
    if (need(a, 2, "state lowpower-set <0|1>")) return 1;
    r = c.LowPowerSwitch(std::atoi(a[1].c_str()));
    printf("LowPowerSwitch(%s) ret=%d\n", a[1].c_str(), r);
  } else if (act == "report") {
    if (need(a, 3, "state report <interval_ms> <dur_ms>")) return 1;
    r = c.SetReportFreq(std::atoi(a[1].c_str()), std::atoi(a[2].c_str()));
    printf("SetReportFreq(%s, %s) ret=%d\n", a[1].c_str(), a[2].c_str(), r);
  } else if (act == "switch") {
    if (need(a, 3, "state switch <name> <0|1>")) return 1;
    int32_t status = 0;
    r = c.ServiceSwitch(a[1], std::atoi(a[2].c_str()), status);
    printf("ServiceSwitch(%s, %s) ret=%d status=%d\n", a[1].c_str(), a[2].c_str(), r, status);
  } else {
    fprintf(stderr, "未知 state action: %s\n", act.c_str());
    return 1;
  }
  return r == 0 ? 0 : 2;
}

// ---------------------------------------------------------------------------
// audio (voice)
// ---------------------------------------------------------------------------
static int cmd_audio(const vector<string>& a) {
  if (a.empty()) { fprintf(stderr, "用法: audio <action>\n"); return 1; }
  unitree::robot::r1::AudioClient c;
  c.Init();
  c.SetTimeout(5.0f);
  const string& act = a[0];
  int r = 0;

  if (act == "tts") {
    if (need(a, 2, "audio tts <text> [speaker]")) return 1;
    const int speaker = (a.size() > 2) ? std::atoi(a[2].c_str()) : 0;
    r = c.TtsMaker(a[1], speaker);
    printf("TtsMaker ret=%d speaker=%d text=\"%s\"\n", r, speaker, a[1].c_str());
  } else if (act == "vol-get") {
    uint8_t v = 0;
    r = c.GetVolume(v);
    printf("GetVolume ret=%d volume=%u\n", r, v);
  } else if (act == "vol-set") {
    if (need(a, 2, "audio vol-set <0-100>")) return 1;
    r = c.SetVolume(static_cast<uint8_t>(std::atoi(a[1].c_str())));
    printf("SetVolume(%s) ret=%d\n", a[1].c_str(), r);
  } else if (act == "led") {
    if (need(a, 4, "audio led <r> <g> <b>")) return 1;
    r = c.LedControl(static_cast<uint8_t>(std::atoi(a[1].c_str())),
                     static_cast<uint8_t>(std::atoi(a[2].c_str())),
                     static_cast<uint8_t>(std::atoi(a[3].c_str())));
    printf("LedControl(%s,%s,%s) ret=%d\n", a[1].c_str(), a[2].c_str(), a[3].c_str(), r);
  } else if (act == "asr") {
    const double secs = (a.size() > 1) ? std::atof(a[1].c_str()) : 10.0;
    printf("监听 ASR (rt/audio_msg) %g 秒...\n", secs);
    unitree::robot::ChannelSubscriber<std_msgs::msg::dds_::String_> sub("rt/audio_msg");
    sub.InitChannel([](const void* msg) {
      auto* m = static_cast<const std_msgs::msg::dds_::String_*>(msg);
      printf("[asr] %s\n", m->data().c_str());
    });
    std::this_thread::sleep_for(std::chrono::duration<double>(secs));
    sub.CloseChannel();
  } else {
    fprintf(stderr, "未知 audio action: %s\n", act.c_str());
    return 1;
  }
  return r == 0 ? 0 : 2;
}

// ---------------------------------------------------------------------------
// config
// ---------------------------------------------------------------------------
static int cmd_config(const vector<string>& a) {
  if (a.empty()) { fprintf(stderr, "用法: config <action>\n"); return 1; }
  unitree::robot::b2::ConfigClient c;
  c.Init();
  c.SetTimeout(5.0f);
  const string& act = a[0];
  int r = 0;

  if (act == "get") {
    if (need(a, 2, "config get <name>")) return 1;
    string content;
    r = c.Get(a[1], content);
    printf("Get(%s) ret=%d content=%s\n", a[1].c_str(), r, content.c_str());
  } else if (act == "set") {
    if (need(a, 3, "config set <name> <content>")) return 1;
    r = c.Set(a[1], a[2]);
    printf("Set(%s) ret=%d\n", a[1].c_str(), r);
  } else if (act == "del") {
    if (need(a, 2, "config del <name>")) return 1;
    r = c.Del(a[1]);
    printf("Del(%s) ret=%d\n", a[1].c_str(), r);
  } else if (act == "meta") {
    if (need(a, 2, "config meta <name>")) return 1;
    string meta;
    r = c.Meta(a[1], meta);
    printf("Meta(%s) ret=%d meta=%s\n", a[1].c_str(), r, meta.c_str());
  } else {
    fprintf(stderr, "未知 config action: %s\n", act.c_str());
    return 1;
  }
  return r == 0 ? 0 : 2;
}

// ---------------------------------------------------------------------------
// lowstate（只读）
// ---------------------------------------------------------------------------
namespace {
std::mutex g_low_mtx;
unitree_hg::msg::dds_::LowState_ g_low;
std::atomic<bool> g_low_have{false};

void lowstate_handler(const void* msg) {
  std::lock_guard<std::mutex> lk(g_low_mtx);
  g_low = *static_cast<const unitree_hg::msg::dds_::LowState_*>(msg);
  g_low_have = true;
}
}  // namespace

static int cmd_lowstate(const vector<string>& a) {
  if (a.empty() || a[0] != "dump") {
    fprintf(stderr, "用法: lowstate dump [sec]\n");
    return 1;
  }
  const double secs = (a.size() > 1) ? std::atof(a[1].c_str()) : 3.0;

  unitree::robot::ChannelSubscriber<unitree_hg::msg::dds_::LowState_> sub("rt/lowstate");
  sub.InitChannel(lowstate_handler);

  printf("等待 rt/lowstate ...\n");
  auto t0 = std::chrono::steady_clock::now();
  while (!g_low_have.load() && std::chrono::steady_clock::now() - t0 < std::chrono::seconds(5)) {
    std::this_thread::sleep_for(std::chrono::milliseconds(50));
  }
  if (!g_low_have.load()) {
    printf("未收到 rt/lowstate（机器人未连接/网卡不对）\n");
    sub.CloseChannel();
    return 2;
  }

  const auto deadline = std::chrono::steady_clock::now() + std::chrono::duration<double>(secs);
  while (std::chrono::steady_clock::now() < deadline) {
    unitree_hg::msg::dds_::LowState_ s;
    { std::lock_guard<std::mutex> lk(g_low_mtx); s = g_low; }
    const auto& imu = s.imu_state();
    printf("---- mode_machine=%u  rpy=[%.3f %.3f %.3f]  gyro=[%.3f %.3f %.3f]  acc=[%.3f %.3f %.3f]\n",
           static_cast<unsigned>(s.mode_machine()),
           imu.rpy()[0], imu.rpy()[1], imu.rpy()[2],
           imu.gyroscope()[0], imu.gyroscope()[1], imu.gyroscope()[2],
           imu.accelerometer()[0], imu.accelerometer()[1], imu.accelerometer()[2]);
    printf("     idx:  q       dq      tau_est   mode\n");
    for (int i = 0; i < 35; ++i) {
      const auto& m = s.motor_state()[i];
      printf("     %2d:  %7.3f %7.3f %7.3f   %u\n", i, m.q(), m.dq(), m.tau_est(),
             static_cast<unsigned>(m.mode()));
    }
    std::this_thread::sleep_for(std::chrono::seconds(1));
  }
  sub.CloseChannel();
  return 0;
}

// ---------------------------------------------------------------------------
int main(int argc, char** argv) {
  if (argc < 2) { print_usage(); return 1; }
  const string first = argv[1];
  if (first == "help" || first == "-h" || first == "--help") { print_usage(); return 0; }

  if (argc < 3) {
    fprintf(stderr, "用法: r1_tool <network_interface> <group> <action> [args...]\n");
    return 1;
  }
  const string iface = argv[1];
  const string group = argv[2];
  vector<string> args(argv + 3, argv + argc);

  unitree::robot::ChannelFactory::Instance()->Init(0, iface);

  if (group == "loco")      return cmd_loco(args);
  if (group == "msc")       return cmd_msc(args);
  if (group == "state")     return cmd_state(args);
  if (group == "audio")     return cmd_audio(args);
  if (group == "config")    return cmd_config(args);
  if (group == "lowstate")  return cmd_lowstate(args);

  fprintf(stderr, "未知 group: %s\n", group.c_str());
  print_usage();
  return 1;
}

File: example/r1/high_level/scripts/r1_tool.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""R1 官方 SDK 全功能交互终端（部署于 R1-EDU 开发计算单元 / 算力背包）。

把所有官方服务都做成菜单，全部指令和参数提示都显示在终端里，无需记忆：

  运动控制 loco       (sport)          FSM、阻尼/站立/启动/零力矩、移动、速度档位
  运动模式 motion_switcher             check/select/release/silent
  机器人状态 robot_state               服务列表、版本、低功耗、上报频率、服务开关
  语音灯光 voice       (audio)          TTS、音量、RGB 灯、ASR
  配置 config                          读/写/删/元信息
  只读状态 lowstate                    订阅 rt/lowstate 打印 IMU + 35 电机

底层调用已编译的 build/bin/r1_tool；不存在时自动调用同目录 build.sh 编译。

用法：
    python3 r1_tool.py                 # 交互菜单
    python3 r1_tool.py --iface eth0
    python3 r1_tool.py loco get-fsm    # 也可直接执行（脚本化）
"""

import argparse
import os
import subprocess
import sys

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.abspath(os.path.join(SCRIPT_DIR, "..", "..", "..", ".."))
TOOL_BIN = os.path.join(ROOT, "build", "bin", "r1_tool")


class Tool:
    def __init__(self, iface=None, yes=False, dry_run=False):
        self.iface = iface
        self.yes = yes
        self.dry_run = dry_run

    def detect_iface(self):
        try:
            out = subprocess.run(["ip", "-o", "-4", "addr", "show"],
                                 capture_output=True, text=True, check=False).stdout
        except FileNotFoundError:
            return None
        for line in out.splitlines():
            p = line.split()
            if len(p) >= 4 and p[3].startswith("192.168.123."):
                return p[1]
        return None

    def ensure_iface(self):
        if not self.iface:
            self.iface = self.detect_iface()
        if not self.iface:
            sys.exit("[r1] 未探测到 192.168.123.x 网卡，请用 --iface <name> 指定（如 --iface eth0）")

    def ensure_bin(self):
        if not (os.path.isfile(TOOL_BIN) and os.access(TOOL_BIN, os.X_OK)):
            print(f"[r1] 未找到 {TOOL_BIN}，正在编译 r1_tool ...")
            subprocess.run([os.path.join(SCRIPT_DIR, "build.sh"), "r1_tool"], check=True)
        if not os.access(TOOL_BIN, os.X_OK):
            sys.exit(f"[r1] 编译后仍未找到 {TOOL_BIN}")

    def confirm(self, msg):
        if self.yes:
            return True
        print(msg)
        try:
            return input("确认执行？输入 YES 继续: ").strip() == "YES"
        except EOFError:
            return False

    def run(self, *args):
        cmd = [TOOL_BIN, self.iface, *args]
        print("[r1] 执行:", " ".join(cmd))
        if self.dry_run:
            return 0
        return subprocess.run(cmd, check=False).returncode


def ask(prompt, default=""):
    try:
        v = input(prompt).strip()
    except EOFError:
        return None
    return v or default


def submenu(tool, title, entries, back="0"):
    """entries: list of (key, label, handler). handler 返回 False 表示返回上级。"""
    while True:
        print(f"\n----------- {title} -----------")
        for key, label, _ in entries:
            print(f"  {key:>3}) {label}")
        print(f"    0) 返回")
        print("-" * (30 + len(title)))
        c = ask("请选择: ")
        if c is None or c == back:
            return
        for key, _label, handler in entries:
            if c == key:
                handler()
                break
        else:
            print("无效选择")


def menu_loco(tool):
    def guard(msg):
        return tool.confirm(msg)

    def set_fsm(fid):
        if fid == 0 and not guard("⚠️  即将切换到【零力矩 ZeroTorque】：机器人会瘫软，请确保有保护！"):
            print("[r1] 已取消"); return
        tool.run("loco", "set-fsm", str(fid))

    def move():
        vx = ask("vx (m/s, 回车=0): ", "0")
        vy = ask("vy (m/s, 回车=0): ", "0")
        w = ask("vyaw (rad/s, 回车=0): ", "0")
        d = ask("duration (s, 回车=1): ", "1")
        if None in (vx, vy, w, d):
            return
        if not guard(f"⚠️  即将以 vx={vx} vy={vy} vyaw={w} 移动 {d}s，请确保周围安全！"):
            print("[r1] 已取消"); return
        tool.run("loco", "move", vx, vy, w, d)

    entries = [
        ("1", "查询 FSM ID / FSM Mode", lambda: (tool.run("loco", "get-fsm"),
                                                 tool.run("loco", "get-fsm-mode"))),
        ("2", "阻尼 Damp            [fsm 1]", lambda: tool.run("loco", "damp")),
        ("3", "站立 StandUp         [fsm 4]", lambda: tool.run("loco", "stand")),
        ("4", "启动主运控 Start     [fsm 811]", lambda: tool.run("loco", "start")),
        ("5", "移动 move vx vy vyaw [duration]", move),
        ("6", "停止移动 StopMove", lambda: tool.run("loco", "stop")),
        ("7", "设置速度档位 SetSpeedMode", lambda: _ask_run(tool, "速度档位:", "loco", "speed")),
        ("8", "零力矩 ZeroTorque    [fsm 0]  ⚠️", lambda: set_fsm(0)),
        ("9", "设置任意 FSM ID", lambda: _ask_run(tool, "FSM ID:", "loco", "set-fsm")),
    ]
    submenu(tool, "运动控制 loco (sport)", entries)


def menu_msc(tool):
    def release():
        if not tool.confirm("⚠️  Release 会释放当前运动模式（如 ai_sport），机器人将失去该服务控制。"):
            print("[r1] 已取消"); return
        tool.run("msc", "release")

    entries = [
        ("1", "查询当前模式 CheckMode", lambda: tool.run("msc", "check")),
        ("2", "选择模式 SelectMode (如 ai)", lambda: _ask_run(tool, "模式名/别名:", "msc", "select")),
        ("3", "释放模式 ReleaseMode  ⚠️", release),
        ("4", "查询静默 GetSilent", lambda: tool.run("msc", "silent-get")),
        ("5", "设置静默 SetSilent <0|1>", lambda: _ask_run(tool, "0/1:", "msc", "silent-set")),
    ]
    submenu(tool, "运动模式 motion_switcher", entries)


def menu_state(tool):
    entries = [
        ("1", "列出所有服务及状态 ServiceList", lambda: tool.run("state", "services")),
        ("2", "查询版本 GetPkgVersion", lambda: tool.run("state", "version")),
        ("3", "低功耗状态 LowPowerStatus", lambda: tool.run("state", "lowpower-get")),
        ("4", "低功耗开关 LowPowerSwitch <0|1>", lambda: _ask_run(tool, "0/1:", "state", "lowpower-set")),
        ("5", "设置上报频率 SetReportFreq <interval> <duration>", lambda: _ask2_run(tool, "interval(ms):", "duration(ms):", "state", "report")),
        ("6", "服务开关 ServiceSwitch <name> <0|1>", lambda: _ask2_run(tool, "服务名:", "0/1:", "state", "switch")),
    ]
    submenu(tool, "机器人状态 robot_state", entries)


def menu_audio(tool):
    entries = [
        ("1", "语音合成 TTS <text> [speaker]", lambda: _tts(tool)),
        ("2", "查询音量 GetVolume", lambda: tool.run("audio", "vol-get")),
        ("3", "设置音量 SetVolume <0-100>", lambda: _ask_run(tool, "音量 0-100:", "audio", "vol-set")),
        ("4", "RGB 灯 LedControl <r> <g> <b>", lambda: _led(tool)),
        ("5", "监听 ASR 识别 rt/audio_msg [sec]", lambda: _ask_run(tool, "监听秒数(回车=10):", "audio", "asr", allow_empty=True)),
    ]
    submenu(tool, "语音/灯 voice", entries)


def menu_config(tool):
    entries = [
        ("1", "读取配置 config get <name>", lambda: _ask_run(tool, "配置名:", "config", "get")),
        ("2", "写配置 config set <name> <content>", lambda: _ask2_run(tool, "配置名:", "内容:", "config", "set")),
        ("3", "删除配置 config del <name>", lambda: _ask_run(tool, "配置名:", "config", "del")),
        ("4", "配置元信息 config meta <name>", lambda: _ask_run(tool, "配置名:", "config", "meta")),
    ]
    submenu(tool, "配置 config", entries)


def menu_lowstate(tool):
    entries = [
        ("1", "打印 IMU + 35 电机状态 (3s)", lambda: tool.run("lowstate", "dump", "3")),
        ("2", "自定义时长 lowstate dump <sec>", lambda: _ask_run(tool, "秒数(回车=3):", "lowstate", "dump", allow_empty=True)),
    ]
    submenu(tool, "只读状态 lowstate", entries)


# ---- 参数辅助 ----
def _ask_run(tool, prompt, group, action, allow_empty=False):
    v = ask(prompt)
    if v is None:
        return
    args = [group, action]
    if v != "":
        args.append(v)
    elif not allow_empty:
        print("[r1] 未输入，取消"); return
    tool.run(*args)


def _ask2_run(tool, p1, p2, group, action):
    a = ask(p1)
    b = ask(p2)
    if a is None or b is None or a == "" or b == "":
        print("[r1] 输入不完整，取消"); return
    tool.run(group, action, a, b)


def _tts(tool):
    text = ask("要合成的文本: ")
    if not text:
        print("[r1] 空文本，取消"); return
    spk = ask("speaker id (回车=0): ", "0")
    tool.run("audio", "tts", text, spk)


def _led(tool):
    r = ask("R (0-255): ")
    g = ask("G (0-255): ")
    b = ask("B (0-255): ")
    if None in (r, g, b) or "" in (r, g, b):
        print("[r1] 输入不完整，取消"); return
    tool.run("audio", "led", r, g, b)


MENUS = [
    ("1", "运动控制 loco (sport)      阻尼/站立/启动/零力矩/移动/速度", menu_loco),
    ("2", "运动模式 motion_switcher   check/select/release/silent", menu_msc),
    ("3", "机器人状态 robot_state     服务列表/版本/低功耗/上报/开关", menu_state),
    ("4", "语音/灯 voice             TTS/音量/RGB/ASR", menu_audio),
    ("5", "配置 config               读/写/删/元信息", menu_config),
    ("6", "只读状态 lowstate         IMU + 35 电机", menu_lowstate),
]


def main_menu(tool):
    print(f"[r1] 网卡: {tool.iface}")
    print(f"[r1] 二进制: {TOOL_BIN}")
    while True:
        print("\n============== R1 官方 SDK 全功能测试 ==============")
        for k, label, _ in MENUS:
            print(f"  {k}) {label}")
        print("  0) 退出")
        print("====================================================")
        c = ask("请选择: ")
        if c is None or c == "0":
            return
        for k, _label, fn in MENUS:
            if c == k:
                fn(tool)
                break
        else:
            print("无效选择")


def main():
    ap = argparse.ArgumentParser(description="R1 官方 SDK 全功能交互终端")
    ap.add_argument("--iface", default=None, help="DDS 网卡（默认自动探测 192.168.123.x）")
    ap.add_argument("-y", "--yes", action="store_true", help="跳过危险操作确认")
    ap.add_argument("--dry-run", action="store_true", help="只打印命令，不执行")
    ap.add_argument("args", nargs="*", help="直接执行: <group> <action> [args...]")
    a = ap.parse_args()

    tool = Tool(a.iface, yes=a.yes, dry_run=a.dry_run)
    tool.ensure_iface()
    tool.ensure_bin()

    if a.args:
        sys.exit(tool.run(*a.args))
    main_menu(tool)


if __name__ == "__main__":
    main()
